src/Halogen.purs

```diff
@@ -1,159 +1,78 @@
-
--- | A variant of `runUI` which supports a _post-render hook_. This allows applications
--- | to support third-party components or other custom behaviors by modifying the DOM after
--- | each update.
--- |
--- | This is considered an advanced feature, and should only be used with an understanding of
--- | the rendering pipeline.
-runUIWith :: forall req eff.
-               Component (Event (HalogenEffects eff)) req req ->
-               (req -> HTMLElement -> Driver req eff -> Eff (HalogenEffects eff) Unit) ->
-               Eff (HalogenEffects eff) (Tuple HTMLElement (Driver req eff))
-runUIWith sf postRender = mainLoop (pure <<< componentProcess sf postRender)
-
--- | A `Process` receives inputs and outputs effectful computations which update the DOM.
-type Process req eff = SF (Tuple req HTMLElement) (Eff (HalogenEffects eff) HTMLElement)
-
--- | Build a `Process` from a `Component`.
-componentProcess :: forall req eff.
-                      Component (Event (HalogenEffects eff)) req req ->
-                      (req -> HTMLElement -> Driver req eff -> Eff (HalogenEffects eff) Unit) ->
-                      Driver req eff ->
-                      Tuple HTMLElement (Process req eff)
-componentProcess sf postRender driver =
-  let render  = R.renderHTML requestHandler
-      vtrees  = render <$> sf
-      diffs   = tail vtrees >>> changes (head vtrees)
-      process = (diffs &&& input) *** input
-      node    = createElement (head vtrees)
-  in Tuple node (applyPatch <$> process)
-  where
-  requestHandler :: Event (HalogenEffects eff) req -> Eff (HalogenEffects eff) Unit
-  requestHandler aff = runEvent logger driver aff
-
-  logger :: Error -> Eff (HalogenEffects eff) Unit
-  logger e = log $ "Uncaught error in asynchronous code: " <> message e
-
-  applyPatch :: Tuple (Tuple Patch req) HTMLElement -> Eff (HalogenEffects eff) HTMLElement
-  applyPatch (Tuple (Tuple p req) node) = do
-    node' <- patch p node
-    postRender req node' driver
-    return node'
+import Control.Monad.Eff (Eff())
+import Control.Monad.Eff.Exception (throwException, error)
+import Control.Monad.Eff.Ref (Ref(), newRef, writeRef, readRef)
+import Control.Monad.Free
+import Control.Monad.State (runState)
+import Control.Monad.State.Trans (runStateT)
+
+import Data.Coyoneda (Coyoneda(), liftCoyoneda)
+import Data.DOM.Simple.Types (HTMLElement())
+import Data.Inject (Inject, inj)
+import Data.Maybe (Maybe(..))
+import Data.Tuple (Tuple(..))
+import Data.Void (Void())
+
+import Halogen.Component (Component(), renderComponent, queryComponent)
+import Halogen.Effects (HalogenEffects())
+import Halogen.HTML.Renderer.VirtualDOM (renderHTML)
+import Halogen.Internal.VirtualDOM (VTree(), createElement, diff, patch)
+
+type DriverState s =
+  Maybe { node :: HTMLElement
+        , vtree :: VTree
+        , state :: s
+        }
+
+type Driver f eff = forall i. f i -> Eff (HalogenEffects eff) i
+
+runUI :: forall eff s f. Component s f (Eff (HalogenEffects eff)) Void
```

Comment:
Most fast promises library use something similar to asap, which is essentially bouncing like what monad rec does. The promises spec only requires that it happens after the current call stack, not that it happens on a subsequent tick. I've modified some hot algorithms in sweet.js to use Bluebird as an expirement, and only saw about a 10-15% decrease in performance. The raw part of asap.js is small enough to just inline as well. It's certainly not any longer than other FFI code in Aff :)